# Detect Phishing URLs
### Capstone 3 - Data Wrangling and EDA
Michael Garber

#### High-Level Steps
1. Data Wrangling
    1. Data Collection
    2. Data Organization
    3. Data Definition
    4. Data Cleaning
1. Exploratory Data Analysis


#### Data Wrangling

##### Data Collection
Info/Data
- https://www.kaggle.com/datasets/hassaanmustafavi/phishing-urls-dataset

In [8]:
# Please Install KaggleHub if you don't already have it
%pip install kagglehub

In [9]:
# Import Packages
import pandas as pd
import kagglehub
import os
import shutil

In [10]:
# Download Data
kPath = kagglehub.dataset_download("hassaanmustafavi/phishing-urls-dataset")

In [14]:
# Move File to project folder
fileName = 'url_dataset.csv'
srcFilePath = os.path.join(kPath, fileName)
dataDir = os.path.join(os.pardir, 'data', 'raw', fileName)
shutil.move(srcFilePath, dataDir)

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\MG-Laptop\\.cache\\kagglehub\\datasets\\hassaanmustafavi\\phishing-urls-dataset\\versions\\1\\url_dataset.csv'

In [16]:
# Create pandas dataframe
urlData = pd.read_csv(dataDir)

In [18]:
# check dataframe - # or rows
print(urlData.shape)

(450176, 2)


##### Data Organization
Project file structure based on the cookiecutter data science template. \
[https://drivendata.github.io/cookiecutter-data-science/](https://drivendata.github.io/cookiecutter-data-science/)

Folder structure tree (GitHub) \
[https://github.com/mdgarber/DetectPhishURL/blob/7dd7d38c001590b4629f8810906a3724ab107fd5/DetectPhishURL/README.md](https://github.com/mdgarber/DetectPhishURL/blob/7dd7d38c001590b4629f8810906a3724ab107fd5/DetectPhishURL/README.md)

##### Data Definition

- Column names
- Data types
- Description of the columns
- Counts and unique values
- Ranges of values
- Calc Summary statistics

In [22]:
# Data types, unique values, range of index
urlData.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450176 entries, 0 to 450175
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   url     450176 non-null  object
 1   type    450176 non-null  object
dtypes: object(2)
memory usage: 6.9+ MB


Description of the columns
- url - The web address to be analyzed.
- type - The classification of the URL (phishing or legitimate).

In [93]:
# Value counts for fields
print('\n======== Counts ========\n')
print(urlData['url' ].value_counts())
print('\n') # line space
print(urlData['type'].value_counts())


======== Counts ========

url
https://www.google.com                                                                                                                             1
https://www.tabheaven.com/gales-eric-tabs.html                                                                                                     1
https://www.tabs-database.com/justin-mcroberts-chords.html                                                                                         1
https://www.tabpower.com/a806.html                                                                                                                 1
https://www.tabor.edu/alumni/directory?decade=1960                                                                                                 1
                                                                                                                                                  ..
https://www.billboard.com/artist/anita-pointer/discography/songs/23169     

In [33]:
# describe dataframe
urlData.describe()

url        type
count                   450176      450176
unique                  450176           2
top     https://www.google.com  legitimate
freq                         1      345738

Data Definition Summary

- Column names
    - url
        - Data types
            - series of string
        - Description of the column 
            - The web address to be analyzed.
        - Counts and unique values 
            - 450176 (all unique)
        - Ranges of values 
            - all unique URLs
    - type
        - Data types 
            - series of string
        - Description of the column
            - The classification of the URL (phishing or legitimate).
        - Counts and unique values
            - legitimate    345738
            - phishing      104438
        - Ranges of values
            - phishing
            - legitimate
            
- Calc Summary statistics
    - N/A ... text data

##### Data Cleaning

- this dataset is already clean  =)

#### EDA

In [85]:
# view the data head
urlData.head(10)

url        type
0     https://www.google.com  legitimate
1    https://www.youtube.com  legitimate
2   https://www.facebook.com  legitimate
3      https://www.baidu.com  legitimate
4  https://www.wikipedia.org  legitimate
5     https://www.reddit.com  legitimate
6      https://www.yahoo.com  legitimate
7   https://www.google.co.in  legitimate
8         https://www.qq.com  legitimate
9     https://www.amazon.com  legitimate

In [89]:
# view the data head - only phish
urlData[urlData['type'] == 'phishing'].head(10)

url      type
345738                   http://atualizacaodedados.online  phishing
345739                     http://webmasteradmin.ukit.me/  phishing
345740           http://stcdxmt.bigperl.in/klxtv/apps/uk/  phishing
345741   https://tubuh-syarikat.com/plugins/fields/files/  phishing
345742  http://rolyborgesmd.com/exceword/excel.php?.ra...  phishing
345743  http://ongelezen-voda.000webhostapp.com/inlogg...  phishing
345744  http://www.valenzaceramic.com/home/webapps/e52...  phishing
345745  http://membership-issue.forteimpex.com/dk2mmm=...  phishing
345746  http://membership-issue.forteimpex.com/dk2mmm=...  phishing
345747  http://chronopost-service-enligne.net/56123s/r...  phishing

In [102]:
# Export df to file
urlData.to_csv('../data/interim/urlData_raw.csv')